---
title: "limma_voom_diff_expression"
author: "Shay Simmons"
date: "1/19/2023"
output: html_document
---

######################## Preparing Work Space
```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

```{r setup, include=FALSE}

suppressPackageStartupMessages({

  library(tidyverse)
  library(data.table)
  library(compositions)
  library(kableExtra)
  library(plyr)
  library(cowplot)
  library(gridExtra)
  library(ggpubr)
  library(limma)
  library(edgeR)
  library(variancePartition)
  library(enrichplot)
  library(clusterProfiler)
}) 

#source("./common_functions/common_functions.R")
colorArray = colors()[grep("grey|gray", colors(), invert=TRUE)]

```


######################## Differential Expression Testing
```{r Normalize Counts and Establish Filters for Lowly Expressed Genes}

# Filter genes - This gene filter is placed at 0.5 counts per million
isexpr.lst = dlply(METADATA, .(Cohort), .fun = function(mtd){
	rowSums(cpm(geneCountsMerged[,mtd$SampleID])>.5) >= .3*ncol(geneCountsMerged[,mtd$SampleID])
	})
isexpr = do.call(rbind, isexpr.lst)

# Keep gene expressed in both cohorts that also fall within the filter
isexpr = colSums(isexpr) == 2 
table(isexpr)
   
# ormalize each cohort separately -- remember this is where limma differs from DESeq2
# Function of the loops
# dge list is keeping the cohorts separate
dge.lst = lapply(levels(METADATA$Cohort), function(Cohort){
 # Add a frame with gene lists
   # normalize counts
  dge = DGEList(geneCountsMerged[isexpr,METADATA$Cohort==Cohort]) # remove the [isexpr,METADATA$Cohort==Cohort]
  calcNormFactors(dge, "none") # TMM
})
names(dge.lst) = levels(METADATA$Cohort)

```

```{r Differential Expression Testing}

form.lst = list(
  `MSSM-Penn-Pitt` = ~ 0 + Group + Sex + scale(RIN) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*(ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3),
  `NIMH-HBCC` = ~ 0 + Group + Sex + scale(RIN) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + scale(ageOfDeath) + scale(cellFrac_ilr_1) + scale(cellFrac_ilr_2) + scale(cellFrac_ilr_3)
  )

# dream analysis
fit.lst = lapply(dge.lst, function(dge){

  i = match(colnames(dge), rownames(METADATA))
  info = METADATA[i,]
  form_use = form.lst[[info$Cohort[1]]]
  
# Runs independently with group 
L = makeContrastsDream(form_use, info,
                         contrasts = c(AA_C = "(GroupAAControl + GroupAASCZ) - (GroupCControl + GroupCSCZ)",
                                       SCZ_control = "(GroupAASCZ + GroupCSCZ) - (GroupAAControl + GroupCControl)",
                                       Interaction = "(GroupCControl - GroupCSCZ - GroupAAControl + GroupAASCZ)/4"))
 

  vobj_dream = voomWithDreamWeights(dge, form_use, info, plot = F)
  fit = dream(vobj_dream, form_use, info, L)
  fit = eBayes(fit)

  })

names(fit.lst) = names(dge.lst)

```

```{r Name the Genes & Merge the Data Frames}

# Get background genes 
backgroundGenes = data.frame(gene_id = rownames(geneCountsMerged)) %>%
  dplyr::mutate(id = gene_id) %>%
  tidyr::separate(id, c('ensembl_gene_id','position'), sep = '\\.')

# Define biomart object
mart = biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                        dataset = "hsapiens_gene_ensembl",
                        host = "https://dec2021.archive.ensembl.org")

# Query biomart
Ensemble2HGNC = biomaRt::getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "description", "percentage_gene_gc_content", "gene_biotype", "chromosome_name"),
                               filters = "ensembl_gene_id", 
                               values = backgroundGenes$ensembl_gene_id,
                               mart = mart)

```


######################## Data Frame Creation
```{r Specify the contrast you want to see}

DE = lapply(fit.lst, function(FIT){
  topTable(FIT, coef = c("AA_C"), number = Inf) %>%
    tibble::rownames_to_column(var = 'gene_id') %>% 
    dplyr::left_join(backgroundGenes) %>% 
    dplyr::left_join(Ensemble2HGNC) 
})

# do not filter genes for the meta-analysis
de_genes = do.call(rbind, DE) %>% 
  # filter(adj.P.Val < 0.05) %>% 
  dplyr::pull(gene_id) %>% unique()

```

```{r Variance explained by each term}

# combine normalized counts 
dge.comb = cbind(dge.lst$`NIMH-HBCC`$counts,
                 dge.lst$`MSSM-Penn-Pitt`$counts)

form = ~ Group + Sex + scale(RIN) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*(ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3)

vobj_dream = voomWithDreamWeights(dge.comb, form, METADATA, plot = F)
vp = fitExtractVarPartModel(vobj_dream, form, METADATA)

#Plots the variance -- takes too long
#variance_fig <- plotVarPart(vp)
#variance_fig
# ggsave("variance_fig.png")
# Plots the bars per gene - this stratifies PER GENE 
#plotPercentBars( vp[1:3,] )

```

```{r Residualize & multiply back the main effects we want}

METADATA$Ethnicity = factor(METADATA$Ethnicity, levels = c("C","AA"))
METADATA$Dx = factor(METADATA$Dx, levels = c("Control","SCZ"))

# within cohorts 
dge.lst = lapply(levels(METADATA$Cohort), function(Cohort){
 # Add a frame with gene lists
   # normalize counts
  dge = DGEList(geneCountsMerged[isexpr,METADATA$Cohort==Cohort]) # remove the [isexpr,METADATA$Cohort==Cohort]
  calcNormFactors(dge, "none") # TMM
})
names(dge.lst) = levels(METADATA$Cohort)

form.lst = list(
  `MSSM-Penn-Pitt` = ~ Dx*Ethnicity + Sex + scale(RIN) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*(ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3),
  `NIMH-HBCC` = ~ Dx*Ethnicity + Sex + scale(RIN) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + scale(ageOfDeath) + scale(cellFrac_ilr_1) + scale(cellFrac_ilr_2) + scale(cellFrac_ilr_3)
  )

vobj.lst = lapply(dge.lst, function(dge){
  i = match(colnames(dge), rownames(METADATA))
  info = METADATA[i,]
  form_use = form.lst[[info$Cohort[1]]]
  voomWithDreamWeights(dge, form_use, info, save.plot=TRUE, BPPARAM = SnowParam(18))
  })
names(vobj.lst) = names(dge.lst)

eval_residuals = function( form ){
  # Compute residuals, but add back in Dx*Reported_Gender
  # for form = ~ Dx*Reported_Gender
  f = function(fit){
    residuals(fit) + get_prediction( fit, form)
  }
  residuals.lst = lapply( vobj.lst, function(vobj){
    i = match(colnames(vobj), rownames(METADATA))
    info = METADATA[i,]
    form_use = form.lst[[info$Cohort[1]]]
    # Compute residuals which adding back Diagnosis and Race effects
    resid.lst = fitVarPartModel(vobj, form_use, info, fxn = f, BPPARAM = SnowParam(18))  
    do.call(rbind, resid.lst)
    })
  names(residuals.lst) = names(vobj.lst)
  residuals.lst
}
resid.add = list( Dx_Ethnicity  = eval_residuals( ~ Dx*Ethnicity ),
                 Dx  = eval_residuals( ~ Dx ),
                 Ethnicity = eval_residuals( ~ Ethnicity ))

```


######################## Dataframe Creation
```{r Create the residual dataframes}

# Data frames of race
residuals_nimh_race = resid.add[3]$Ethnicity$`NIMH-HBCC`
  residuals_ppm_race = resid.add[3]$Ethnicity$`MSSM-Penn-Pitt`

# Data frames of dx
residuals_nimh_dx = resid.add[2]$Dx$`NIMH-HBCC`
  residuals_ppm_dx = resid.add[2]$Dx$`MSSM-Penn-Pitt`
  
# Data frames of interaction
residuals_nimh_interaction = resid.add[1]$Dx_Ethnicity$`NIMH-HBCC`
  residuals_ppm_interaction = resid.add[1]$Dx_Ethnicity$`MSSM-Penn-Pitt`
  
```

```{r Save the data frames to my directory}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Create the csv files -- Race
#write.csv(residuals_nimh_race, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Race\\residuals_nimh_race.csv",row.names=TRUE)

#write.csv(residuals_ppm_race, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Race\\residuals_ppm_race.csv", row.names=TRUE)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Create the csv files -- Dx
#write.csv(residuals_nimh_dx, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Dx\\residuals_nimh_dx.csv", row.names=TRUE)

#write.csv(residuals_ppm_dx, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Dx\\residuals_ppm_dx.csv", row.names=TRUE)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Create the csv files -- Interaction
#write.csv(residuals_nimh_interaction, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Interaction\\residuals_nimh_interaction.csv", row.names=TRUE)

#write.csv(residuals_ppm_interaction, "/Users/shuaybsimmons/Desktop/School/UofT MSc Medical Science/Research/Thesis Research/Data Frames/MEGENA residual df /Interaction\\residuals_ppm_interaction.csv", row.names=TRUE)



```

